packages/web3torrent/src/library/web3torrent-lib.ts

```diff
@@ -367,85 +370,86 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
     });
 
     torrent.on(TorrentEvents.NOTICE, async (wire, {command, data}) => {
-      log(`< ${command} received from ${wire.peerExtendedHandshake.pseAccount}`, data);
-      let message: Message;
       switch (command) {
         case PaidStreamingExtensionNotices.STOP: // synonymous with a prompt for a payment
           if (torrent.paused) {
-            // We currently treat pausing torrent as canceling downloads
             await this.closeDownloadingChannels(torrent);
           } else if (!torrent.done) {
-            const channelId = data;
-            await this.paymentChannelClient.makePayment(
-              channelId,
-              WEI_PER_BYTE.mul(BUFFER_REFILL_RATE).toString()
-            );
-
-            let balance: string;
-            const channel = this.paymentChannelClient.channelCache[channelId];
-
-            if (channel) {
-              balance = channel.beneficiaryBalance;
-            } else {
-              balance = 'unknown';
-            }
-
-            log(
-              `attempted to make payment for channel ${channelId}, beneficiaryBalance: ${balance}`
-            );
+            await this.makePayment(torrent, wire);
           }
           break;
         case PaidStreamingExtensionNotices.START:
           this.jumpStart(torrent, wire);
           break;
         case PaidStreamingExtensionNotices.MESSAGE:
-          message = JSON.parse(data.message);
-          if (message.recipient === this.pseAccount) {
-            await this.paymentChannelClient.pushMessage(message);
-          }
-          break;
-        default:
+          await this.paymentChannelClient.pushMessage(data);
           break;
       }
       this.emit(ClientEvents.TORRENT_NOTICE, {torrent, wire, command, data});
     });
 
     torrent.on(TorrentEvents.DONE, async () => {
-      log('Torrent DONE!');
+      log('Torrent DONE!', torrent);
       this.emit(ClientEvents.TORRENT_DONE, {torrent});
       await this.closeDownloadingChannels(torrent);
     });
 
     torrent.on(TorrentEvents.ERROR, err => {
-      log('ERROR: > ', err);
+      log('Torrent ERROR: > ', torrent, err);
       this.emit(ClientEvents.TORRENT_ERROR, {torrent, err});
     });
     torrent.usingPaidStreaming = true;
 
     return torrent;
   }
 
+  /**
+   * Define the amount to pay, and makes the payment
+   */
+  protected async makePayment(torrent: PaidStreamingTorrent, wire: PaidStreamingWire) {
+    const {peerChannelId} = wire.paidStreamingExtension;
+    let amountToPay = WEI_PER_BYTE.mul(BUFFER_REFILL_RATE).toString();
+    if (torrent.downloaded === 0) {
```

Comment:
Perhaps add a comment here that the last piece is always downloaded first.

Is this *always* the case?

Reply:
Sure, I'll add it.

As far as I tested it, yes. The last piece is always requested first. I should test in 1gb+ torrents to make sure, but I observed that behaviour in all my test files (60kb, 120kb, 1.1mb, 3mb 9mb and 33mb)